Add PrimStorable for deriving getting a Storable instance from a Prim instance

andrewthad opened the pull request:
Resolves https://github.com/haskell/primitive/issues/130

My preference is actually to let this sit for a while to make sure that the final implementation of `DerivingVia` actually matches what is suggested in the documentation this PR provides.

RyanGlScott commented on Data/Primitive/Types.hs:
```diff
@@ -130,6 +135,28 @@ sizeOf x = I# (sizeOf# x)
 
+-- | Newtype that uses a 'Prim' instance to give rise to a 'Storable' instance.
+-- This type is intended to be used with the @DerivingVia@ extension available
+-- in GHC 8.6 and up. For example, consider a user-defined 'Prim' instance for
```
> in GHC 8.6 and up.

I truly hope so :)

RyanGlScott commented on Data/Primitive/Types.hs:
```diff
@@ -130,6 +135,28 @@ sizeOf x = I# (sizeOf# x)
+-- a multi-word data type.
+--
+-- > data Uuid = Uuid Word64 Word64
+-- >   deriving Storable via (PrimStorable Statistic)
```
What is `Statistic`?

andrewthad commented:
@RyanGlScott I've fixed the messed up example type name in the docs. Also, I've added some commented out code to the test suite. If you could uncomment it, turn on deriving strategies and deriving via, and run it, that would make me confident that this is correct.

RyanGlScott commented:
With the following changes:

```diff
diff --git a/test/main.hs b/test/main.hs
index 44552d2..1085167 100644
--- a/test/main.hs
+++ b/test/main.hs
@@ -5,6 +5,13 @@
 {-# LANGUAGE UnboxedTuples #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 
+#if __GLASGOW_HASKELL__ >= 805
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE TypeInType #-}
+#endif
+
 import Control.Applicative
 import Control.Monad
 import Control.Monad.Fix (fix)
@@ -26,6 +33,10 @@ import Data.Function (on)
 #if MIN_VERSION_base(4,9,0)
 import Data.Semigroup (stimes)
 #endif
+#if __GLASGOW_HASKELL__ >= 805
+import Data.Kind
+import Foreign.Storable (Storable)
+#endif
 
 import Test.Tasty (defaultMain,testGroup,TestTree)
 import Test.QuickCheck (Arbitrary,Arbitrary1,Gen,(===))
@@ -119,9 +130,11 @@ main = do
       , TQC.testProperty "mapMaybePrimArrayP" (QCCL.mapMaybeMProp int16 int32 mapMaybePrimArrayP)
 #endif
       ]
-    -- , testGroup "PrimStorable"
-    --   [ lawsToTest (QCC.storableLaws (Proxy :: Proxy Derived))
-    --   ]
+#if __GLASGOW_HASKELL__ >= 805
+    , testGroup "PrimStorable"
+      [ lawsToTest (QCC.storableLaws (Proxy :: Proxy Derived))
+      ]
+#endif
     ]
 
 int16 :: Proxy Int16
@@ -180,7 +193,7 @@ intsLessThan :: Int -> [Int]
 intsLessThan i = if i < 1
   then []
   else (i - 1) : intsLessThan (i - 1)
-  
+
 byteArrayCompareProp :: QC.Property
 byteArrayCompareProp = QC.property $ \(xs :: [Word8]) (ys :: [Word8]) ->
   compareLengthFirst xs ys === compare (byteArrayFromList xs) (byteArrayFromList ys)
@@ -279,12 +292,9 @@ iforM_ xs0 f = go 0 xs0 where
   go !_ [] = return ()
   go !ix (x : xs) = f ix x >> go (ix + 1) xs
 
--- TODO: Uncomment this out when GHC 8.6 is release. Also, uncomment
--- the corresponding PrimStorable test group above.
---
--- newtype Derived = Derived Int16
---   deriving newtype (Prim)
---   deriving Storable via (PrimStorable Derived)
-
-
-
+#if __GLASGOW_HASKELL__ >= 805
+newtype Derived = Derived Int16
+  deriving stock (Eq, Show)
+  deriving newtype (Arbitrary, Prim)
+  deriving Storable via (PrimStorable Derived)
+#endif
```

The tests build and pass.

andrewthad commented:
Awesome! What are you thinking in terms of the plan going forward?  Do you want to go ahead and merge this now and apply your patch  to the test suite once DervingVia lands in GHC head?

treeowl commented:
That sounds reasonable to me. The tool is useful even without DerivingVia;
it just gets more useful with it.

On Thu, Apr 19, 2018 at 11:03 AM, Andrew Martin <notifications@github.com>
wrote:

> Awesome! What are you thinking in terms of the plan going forward? Do you
> want to go ahead and merge this now and apply your patch to the test suite
> once DervingVia lands in GHC head?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/primitive/pull/136#issuecomment-382770141>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABzi_WnX8LI3JmsZgq7XezQvkRDQepmXks5tqKc0gaJpZM4Tb3_U>
> .
>

RyanGlScott commented:
Agreed.

cartazio commented:
How does it handle alignment ?

I’m just catching up

On Thu, Apr 19, 2018 at 11:04 AM Ryan Scott <notifications@github.com>
wrote:

> Merged #136 <https://github.com/haskell/primitive/pull/136>.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/primitive/pull/136#event-1584063195>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAQwhbyToiy42CJtOrXVmAnzrxJxZcTks5tqKeSgaJpZM4Tb3_U>
> .
>

andrewthad commented:
The storable instance has the same alignment as this prim instance. This is discussed a little more in [this comment](https://github.com/haskell/primitive/issues/130#issuecomment-380197224).
